privilege/privileges/cache.go

```diff
@@ -270,3 +274,138 @@ func decodeSetToPrivilege(s types.Set) (mysql.PrivilegeType, error) {
 	}
 	return ret, nil
 }
+
+func (record *userRecord) match(user, host string) bool {
+	return record.User == user && patternMatch(record.Host, host)
+}
+
+func (record *dbRecord) match(user, host, db string) bool {
+	return record.User == user && record.DB == db &&
+		patternMatch(record.Host, host)
+}
+
+func (record *tablesPrivRecord) match(user, host, db, table string) bool {
+	return record.User == user && record.DB == db &&
+		record.TableName == table && patternMatch(record.Host, host)
+}
+
+func (record *columnsPrivRecord) match(user, host, db, table, col string) bool {
+	return record.User == user && record.DB == db &&
+		record.TableName == table && record.ColumnName == col &&
+		patternMatch(record.Host, host)
+}
+
+func patternMatch(pattern, str string) bool {
+	for i := 0; i < len(pattern); i++ {
+		p := pattern[i]
+		if p == '%' {
+			return true
+		}
+		if i >= len(str) || p != str[i] {
+			return false
+		}
+	}
+	return len(pattern) == len(str)
+}
+
+// ConnectionVerification verifies the connection have access to TiDB server.
+func (p *MySQLPrivilege) ConnectionVerification(user, host string) bool {
+	for _, record := range p.User {
+		if record.match(user, host) {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *MySQLPrivilege) matchUser(user, host string) *userRecord {
+	for i := 0; i < len(p.User); i++ {
+		record := &p.User[i]
+		if record.match(user, host) {
+			return record
+		}
+	}
+	return nil
+}
+
+func (p *MySQLPrivilege) matchDB(user, host, db string) *dbRecord {
+	for i := 0; i < len(p.DB); i++ {
+		record := &p.DB[i]
+		if record.match(user, host, db) {
+			return record
+		}
+	}
+	return nil
+}
+
+func (p *MySQLPrivilege) matchTables(user, host, db, table string) *tablesPrivRecord {
+	for i := 0; i < len(p.TablesPriv); i++ {
+		record := &p.TablesPriv[i]
+		if record.match(user, host, db, table) {
+			return record
+		}
+	}
+	return nil
+}
+
+func (p *MySQLPrivilege) matchColumns(user, host, db, table, column string) *columnsPrivRecord {
+	for i := 0; i < len(p.ColumnsPriv); i++ {
+		record := &p.ColumnsPriv[i]
+		if record.match(user, host, db, table, column) {
+			return record
+		}
+	}
+	return nil
+}
+
+// RequestVerification checks whether the user have sufficient privileges to do the operation.
+func (p *MySQLPrivilege) RequestVerification(user, host, db, table, column string, priv mysql.PrivilegeType) bool {
+	record1 := p.matchUser(user, host)
+	if record1 != nil && record1.Privileges&priv > 0 {
+		return true
+	}
+
+	record2 := p.matchDB(user, host, db)
+	if record2 != nil && record2.Privileges&priv > 0 {
+		return true
+	}
+
+	record3 := p.matchTables(user, host, db, table)
```

Comment:
Yes, when we are drop database, we don't need check table privs. 
Under this case, `matchTables` would fail and we'll get the right result, just do some unnecessary computation.

We can leave the optimization to future commits, when the code is actually used.